Add mujoco recipe

traversaro opened the pull request:
MuJoCo is a widely used multibody dynamics simulator used in robotics and reinforcement learning. It has been open-sourced on 2022/05/23 : https://github.com/deepmind/mujoco/releases/tag/2.2.0 .

<!--
Thank you very much for putting in this recipe PR!

This repository is very active, so if you need help with a PR, please let the
right people know. There are language-specific teams for reviewing recipes.

| Language        | Name of review team           |
| --------------- | ----------------------------- |
| python          | `@conda-forge/help-python`    |
| python/c hybrid | `@conda-forge/help-python-c`  |
| r               | `@conda-forge/help-r`         |
| java            | `@conda-forge/help-java`      |
| nodejs          | `@conda-forge/help-nodejs`    |
| c/c++           | `@conda-forge/help-c-cpp`     |
| perl            | `@conda-forge/help-perl`      |
| Julia           | `@conda-forge/help-julia`     |
| ruby            | `@conda-forge/help-ruby`      |
| other           | `@conda-forge/staged-recipes` |

Once the PR is ready for review, please mention one of the teams above in a
new comment. i.e. `@conda-forge/help-some-language, ready for review!`
Then, a bot will label the PR as 'review-requested'.

Due to GitHub limitations, first time contributors to conda-forge are unable
to ping conda-forge teams directly, but you can [ask a bot to ping the team][1]
using a special command in a comment on the PR to get the attention of the
`staged-recipes` team. You can also consider asking on our [Gitter channel][2]
if your recipe isn't reviewed promptly.

[1]: https://conda-forge.org/docs/maintainer/infrastructure.html#conda-forge-admin-please-ping-team
[2]: https://gitter.im/conda-forge/conda-forge.github.io
-->

Checklist
- [x] Title of this PR is meaningful: e.g. "Adding my_nifty_package", not "updated meta.yaml".
- [x] License file is packaged (see [here](https://github.com/conda-forge/staged-recipes/blob/5eddbd7fc9d1502169089da06c3688d9759be978/recipes/example/meta.yaml#L64-L73) for an example).
- [x] Source is from official source.
- [ ] Package does not vendor other packages. (If a package uses the source of another package, they should be separate packages or the licenses of all packages need to be packaged).
- [x] If static libraries are linked in, the license of the static library is packaged.
- [x] Build number is 0.
- [x] A tarball (`url`) rather than a repo (e.g. `git_url`) is used in your recipe (see [here](https://conda-forge.org/docs/maintainer/adding_pkgs.html#build-from-tarballs-not-repos) for more details).
- [x] GitHub users listed in the maintainer section have posted a comment confirming they are willing to be listed there.
- [x] When in trouble, please check our [knowledge base documentation](https://conda-forge.org/docs/maintainer/knowledge_base.html) before pinging a team.

conda-forge-linter commented:
Hi! This is the friendly automated conda-forge-linting service.

I just wanted to let you know that I linted all conda-recipes in your PR (```recipes/mujoco```) and found it was in an excellent condition.

bollwyvl commented:
Unfortunately don't have a lot to add right now: 

- on linux:
  - it does mention upstream that they are building with `clang` on linux, not sure if we can force that with e.g. 
- on osx:
  ```
  FAILED: test/CMakeFiles/fixture.dir/fixture.cc.o 
  $BUILD_PREFIX/bin/x86_64-apple-darwin13.4.0-clang++ -DEIGEN_MPL2_ONLY -DMJSTATIC -I$SRC_DIR/test/.. -I$SRC_DIR/include -I$SRC_DIR/test/gmock -I$SRC_DIR/build/_deps/abseil-cpp-src -isystem $SRC_DIR/build/_deps/googletest-src/googletest/include -isystem $SRC_DIR/build/_deps/googletest-src/googletest -isystem $SRC_DIR/build/_deps/googletest-src/googlemock/include -isystem $SRC_DIR/build/_deps/googletest-src/googlemock -march=core2 -mtune=haswell -mssse3 -ftree-vectorize -fPIC -fPIE -fstack-protector-strong -O2 -pipe -stdlib=libc++ -fvisibility-inlines-hidden -fmessage-length=0 -isystem $PREFIX/include -fdebug-prefix-map=$SRC_DIR=/usr/local/src/conda/mujoco-2.2.0 -fdebug-prefix-map=$PREFIX=/usr/local/src/conda-prefix -O3 -DNDEBUG -isysroot /Applications/Xcode_12.4.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.9.sdk -mmacosx-version-min=10.9 -fPIC -fvisibility=hidden -fvisibility-inlines-hidden -fdata-sections -ffunction-sections -std=c++17 -MD -MT test/CMakeFiles/fixture.dir/fixture.cc.o -MF test/CMakeFiles/fixture.dir/fixture.cc.o.d -o test/CMakeFiles/fixture.dir/fixture.cc.o -c $SRC_DIR/test/fixture.cc
  $SRC_DIR/test/fixture.cc:84:7: error: 'open' is unavailable: introduced in macOS 10.15 - see https://conda-forge.org/docs/maintainer/knowledge_base.html#newer-c-features-with-old-sdk
  ifs.open(path, std::ifstream::in);
  ```
  - maybe just needs the [osx version/sdk set](https://conda-forge.org/docs/maintainer/knowledge_base.html#requiring-newer-macos-sdks)
- on windows
  - ???
  - maybe keep that as a stretch goal after the the feedstock is created?

traversaro commented:
> on osx: 
> maybe just needs the [osx version/sdk set](https://conda-forge.org/docs/maintainer/knowledge_base.html#requiring-newer-macos-sdks)

Yes! Back in time this was not supported in staged-recipes, so we can just skip and enable it once the feedstock is available.

> on windows 
> * maybe keep that as a stretch goal after the the feedstock is created?

Good idea.

traversaro commented:
> * on linux:
>       
>       * it does mention upstream that they are building with `clang` on linux, not sure if we can force that with e.g.

I compiled it locally with apt with gcc, so I suspect that is not the problem.

traversaro commented:
> > * on linux:
> >   ```
> >   * it does mention upstream that they are building with `clang` on linux, not sure if we can force that with e.g.
> >   ```
> 
> I compiled it locally with apt with gcc, so I suspect that is not the problem.

I was wrong: the compilation works fine with https://github.com/deepmind/mujoco/commit/70b19d5fd6ddd037a15196474945811a8016aa26 (i.e. current master branch) with both apt's and conda-forge's gcc, but if I switch to `v2.0.0` I indeed get the same compilation error of CI. Probably we can backport the fix that is in master.

traversaro commented:
The history is a bit confusing as 2.2.0 is not a tag in main's history, and the "Initial open sourcing of MuJoCo." commit is different between main and 2.2.0 : 
* main : https://github.com/deepmind/mujoco/commit/1913a02b40703a00a63d98ad14cfe7676cc941fc
* 2.2.0 : https://github.com/deepmind/mujoco/commit/7ef3f3e441d754ef3c2defb3758739f5070f9b3e

However, by manually running a diff between in `main` and `2.2.0` this is the result:
~~~
traversaro@IITICUBLAP257:~/mujoco/build$ diff working.cc not_working.cc
69,70c69
< mjModel* LoadModelFromString(std::string_view xml, char* error,
<                              int error_size) {
---
> mjModel* LoadModelFromString(std::string_view xml, char* error, int error_size) {
83c82
< const std::string GetFileContents(const char* path) {
---
> const std::string GetFileContents(std::string_view path) {
~~~

So https://github.com/conda-forge/staged-recipes/pull/19049/commits/0da0b39d93bdc8f2a0387ebf7fe3565eb1abf652 should fix the problem.

traversaro commented:
Another fix that was in `main` but not in `2.2.0`: https://github.com/conda-forge/staged-recipes/pull/19049/commits/c6a893a35f34e799522e8be3509a9aa0c1a7bf09 .

traversaro commented:
The recipe tests are now failing due to some additional differences between 2.2.0 and main. I will fix this by backporting https://github.com/deepmind/mujoco/commit/098b125d02d6115578a795d1da0f17811df4cf53, but in this case I opened an issue upstream: https://github.com/deepmind/mujoco/issues/296 .

traversaro commented:
Windows build are failing with:
~~~
2022-05-24T08:51:43.9790981Z [18/248] Building C object CMakeFiles\mujoco.dir\src\engine\engine_support.c.obj
2022-05-24T08:51:43.9795122Z FAILED: CMakeFiles/mujoco.dir/src/engine/engine_support.c.obj 
2022-05-24T08:51:44.4202920Z C:\PROGRA~2\MICROS~1\2019\ENTERP~1\VC\Tools\MSVC\1416~1.270\bin\HostX64\x64\cl.exe  /nologo -DCCD_STATIC_DEFINE -DEIGEN_MPL2_ONLY -DMUJOCO_DLL_EXPORTS -D_CRT_SECURE_NO_WARNINGS -D_GNU_SOURCE -Dmujoco_EXPORTS -I%SRC_DIR%\include -I%SRC_DIR%\src -I%SRC_DIR%\build\_deps\ccd-src\src -I%SRC_DIR%\build\_deps\ccd-build\src -I%SRC_DIR%\build\_deps\lodepng-src -I%SRC_DIR%\build\_deps\qhull-src\src\libqhull_r -I%SRC_DIR%\build\_deps\tinyobjloader-src -I%SRC_DIR%\build\_deps\tinyxml2-src /DWIN32 /D_WINDOWS /O2 /Ob2 /DNDEBUG -MD /Gy /Gw /Oi /showIncludes /FoCMakeFiles\mujoco.dir\src\engine\engine_support.c.obj /FdCMakeFiles\mujoco.dir\ /FS -c %SRC_DIR%\src\engine\engine_support.c
2022-05-24T08:51:44.4207918Z %SRC_DIR%\src\engine\engine_support.c(1306): error C2059: syntax error: 'type'
2022-05-24T08:57:00.0404993Z [19/248] Building C object CMakeFiles\mujoco.dir\src\engine\engine_solver.c.obj
2022-05-24T08:57:00.0411736Z [20/248] Building C object CMakeFiles\mujoco.dir\src\engine\engine_print.c.obj
2022-05-24T08:57:00.0423667Z ninja: build stopped: subcommand failed.
2022-05-24T08:57:02.3611243Z Traceback (most recent call last):
2022-05-24T08:57:02.3612844Z   File "D:\a\1\s\.ci_support\build_all.py", line 205, in <module>
2022-05-24T08:57:02.3613853Z     build_all(os.path.join(root_dir, "recipes"), args.arch)
2022-05-24T08:57:02.3614663Z   File "D:\a\1\s\.ci_support\build_all.py", line 105, in build_all
2022-05-24T08:57:02.3615695Z     build_folders(recipes_dir, folders, arch, channel_urls)
2022-05-24T08:57:02.3616472Z   File "D:\a\1\s\.ci_support\build_all.py", line 164, in build_folders
2022-05-24T08:57:02.3617310Z     conda_build.api.build([recipe], config=get_config(arch, channel_urls))
2022-05-24T08:57:02.3618053Z   File "C:\Miniconda\lib\site-packages\conda_build\api.py", line 186, in build
2022-05-24T08:57:02.3618922Z     return build_tree(
2022-05-24T08:57:02.3619895Z   File "C:\Miniconda\lib\site-packages\conda_build\build.py", line 3088, in build_tree
2022-05-24T08:57:02.3627438Z     packages_from_this = build(metadata, stats,
2022-05-24T08:57:02.3628257Z   File "C:\Miniconda\lib\site-packages\conda_build\build.py", line 2179, in build
2022-05-24T08:57:02.3633330Z     windows.build(m, build_file, stats=build_stats, provision_only=provision_only)
2022-05-24T08:57:02.3634159Z   File "C:\Miniconda\lib\site-packages\conda_build\windows.py", line 297, in build
2022-05-24T08:57:02.3635422Z     check_call_env(cmd, cwd=m.config.work_dir, stats=stats, rewrite_stdout_env=rewrite_env)
2022-05-24T08:57:02.3636300Z   File "C:\Miniconda\lib\site-packages\conda_build\utils.py", line 410, in check_call_env
2022-05-24T08:57:02.3637404Z     return _func_defaulting_env_to_os_environ('call', *popenargs, **kwargs)
2022-05-24T08:57:02.3638497Z   File "C:\Miniconda\lib\site-packages\conda_build\utils.py", line 390, in _func_defaulting_env_to_os_environ
2022-05-24T08:57:02.3639455Z     raise subprocess.CalledProcessError(proc.returncode, _args)
2022-05-24T08:57:02.3640316Z subprocess.CalledProcessError: Command '['cmd.exe', '/d', '/c', 'conda_build.bat']' returned non-zero exit status 1.
2022-05-24T08:57:07.4104742Z ##[error]Cmd.exe exited with code '1'.
2022-05-24T08:57:07.4539832Z ##[section]Finishing: Build recipe
~~~
I tried to fix this by switching to clang-cl .

traversaro commented:
> I tried to fix this by switching to clang-cl .

This worked. So macOS and Linux C++ works fine, so we can start the missing points:
* [ ] De-vendoring
* [ ] Python bindings

saran-t commented:
I've just moved the [deepmind/mujoco@2.2.0](https://github.com/deepmind/mujoco/tree/2.2.0) tag to https://github.com/deepmind/mujoco/commit/098b125d02d6115578a795d1da0f17811df4cf53 so the source code archive (mujoco-2.2.0.zip and mujoco-2.2.0.tar.gz) should now build and install correctly. Sorry about the confusion.

traversaro commented:
> I've just moved the [deepmind/mujoco@2.2.0](https://github.com/deepmind/mujoco/tree/2.2.0) tag to [deepmind/mujoco@098b125](https://github.com/deepmind/mujoco/commit/098b125d02d6115578a795d1da0f17811df4cf53) so the source code archive (mujoco-2.2.0.zip and mujoco-2.2.0.tar.gz) should now build and install correctly. Sorry about the confusion.

Thanks @saran-t !

saran-t commented on recipes/mujoco/build_cxx.sh:
```diff
@@ -0,0 +1,21 @@
+      -DMUJOCO_BUILD_TESTS:BOOL=ON \
+      -DMUJOCO_BUILD_EXAMPLES:BOOL=OFF \
+      -DMUJOCO_ENABLE_AVX:BOOL=OFF \
+      -DMUJOCO_ENABLE_AVX_INTRINSICS:BOOL=OFF \
```
Please consider adding `-DCMAKE_INTERPRODEDURAL_OPTIMIZATION:BOOL=ON` as well.

saran-t replied:
typo: should be `-DCMAKE_INTERPROCEDURAL_OPTIMIZATION`

traversaro replied:
Thanks, done in https://github.com/traversaro/staged-recipes/commit/7575534d1b7cd3542967c0d21e615c93c8221222 and https://github.com/traversaro/staged-recipes/commit/a5957f4325fd0a6ff97175e3f67837e4aea40f04 .

traversaro commented:
Latest failures are due to the use of `-Werror` in python bindings and probably we are using a compiled different from the one used to develop the bindings. That should be fixed by https://github.com/conda-forge/staged-recipes/pull/19049/commits/6933151c6165016ef2ad9aadb7e2776b696b8884 .

traversaro commented:
Latest error (reduced):
~~~
2022-05-25T07:03:14.4693022Z       [ 91%] Building CXX object CMakeFiles/_structs.dir/indexers.cc.o
2022-05-25T07:03:14.4699871Z       /home/conda/staged-recipes/build_artifacts/mujoco_1653461546575/_build_env/bin/x86_64-conda-linux-gnu-c++ -DEIGEN_MPL2_ONLY -D_structs_EXPORTS -I/tmp/pip-req-build-2zsasblr/build/temp.linux-x86_64-cpython-39/_deps/abseil-cpp-src -I/mujoco -I/tmp/pip-req-build-2zsasblr/build/temp.linux-x86_64-cpython-39/_deps/eigen-src -isystem /home/conda/staged-recipes/build_artifacts/mujoco_1653461546575/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_plac/include/python3.9 -isystem /tmp/pip-req-build-2zsasblr/build/temp.linux-x86_64-cpython-39/_deps/pybind11-src/include -fvisibility-inlines-hidden -std=c++17 -fmessage-length=0 -march=nocona -mtune=haswell -ftree-vectorize -fPIC -fstack-protector-strong -fno-plt -O2 -ffunction-sections -pipe -isystem /home/conda/staged-recipes/build_artifacts/mujoco_1653461546575/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_plac/include -fdebug-prefix-map=/home/conda/staged-recipes/build_artifacts/mujoco_1653461546575/work=/usr/local/src/conda/mujoco-python-2.2.0 -fdebug-prefix-map=/home/conda/staged-recipes/build_artifacts/mujoco_1653461546575/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_plac=/usr/local/src/conda-prefix -O3 -DNDEBUG -fPIC -fvisibility=hidden -fvisibility-inlines-hidden -fdata-sections -ffunction-sections -Wall -Wno-int-in-bool-context -Wno-maybe-uninitialized -Wno-sign-compare -Wno-stringop-overflow -Wno-stringop-truncation -std=c++17 -MD -MT CMakeFiles/_structs.dir/indexers.cc.o -MF CMakeFiles/_structs.dir/indexers.cc.o.d -o CMakeFiles/_structs.dir/indexers.cc.o -c /tmp/pip-req-build-2zsasblr/mujoco/indexers.cc
2022-05-25T07:03:14.4705468Z       /tmp/pip-req-build-2zsasblr/mujoco/indexers.cc: In member function 'pybind11::array_t<int> mujoco::python::MjModelActuatorViews::trntype()':
2022-05-25T07:03:14.4706576Z       /tmp/pip-req-build-2zsasblr/mujoco/indexers.cc:242:1: error: '#pragma' is not allowed here
2022-05-25T07:03:14.4707163Z         242 | MJMODEL_ACTUATOR
2022-05-25T07:03:14.4707545Z             | ^~~
2022-05-25T07:03:14.4708472Z       /tmp/pip-req-build-2zsasblr/mujoco/indexers.cc: In member function 'pybind11::array_t<int> mujoco::python::MjModelActuatorViews::dyntype()':
2022-05-25T07:03:14.4709752Z       /tmp/pip-req-build-2zsasblr/mujoco/indexers.cc:242:1: error: '#pragma' is not allowed here
2022-05-25T07:03:14.4710908Z       /tmp/pip-req-build-2zsasblr/mujoco/indexers.cc: In member function 'pybind11::array_t<int> mujoco::python::MjModelActuatorViews::gaintype()':
2022-05-25T07:03:14.4711911Z       /tmp/pip-req-build-2zsasblr/mujoco/indexers.cc:242:1: error: '#pragma' is not allowed here
2022-05-25T07:03:14.4712967Z       /tmp/pip-req-build-2zsasblr/mujoco/indexers.cc: In member function 'pybind11::array_t<int> mujoco::python::MjModelActuatorViews::biastype()':
2022-05-25T07:03:14.4713963Z       /tmp/pip-req-build-2zsasblr/mujoco/indexers.cc:242:1: error: '#pragma' is not allowed here
~~~

traversaro commented:
Given the usage of `_Pragma` in the codebase (see https://github.com/deepmind/mujoco/blob/64bc6d27b242d5c5683b4a03dd36f1844b910742/python/mujoco/util/crossplatform.h#L54) and this bug report for GCC: https://gcc.gnu.org/bugzilla/show_bug.cgi?id=78657 this *may* be related to the use of GCC vs Clang, as originally suggested in https://github.com/conda-forge/staged-recipes/pull/19049#issuecomment-1134692637 .

traversaro commented:
New error:
~~~
ImportError: /home/conda/staged-recipes/build_artifacts/mujoco_1653468396456/_test_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold/lib/python3.7/site-packages/mujoco/_enums.cpython-37m-x86_64-linux-gnu.so: undefined symbol: _Znwm
~~~

Possibly related to the fact that mujoco C++ is compiled with GCC and the bindingw with Python?

traversaro commented:
Patches from AUR related to de-vendoring:
* https://aur.archlinux.org/cgit/aur.git/tree/mujoco.patch?h=python-mujoco
* https://aur.archlinux.org/cgit/aur.git/tree/mujoco.patch?h=mujoco

traversaro commented:
> The crash seems to be related to `pyglfw`/`glfw`, so it could be related to the **two** copies of glfw vendored by mujoco and by mujoco-python. So it could make sense to first try to tackle **de-vendoring** `glfw`, and see if the issue is fixed.

Actually no copy of `glfw`  is vendored. Probably the crash is due to the fact that the test is running in an headless system?

traversaro commented:
Windows is failing with:
~~~
[53/53] Linking CXX shared library bin\mujoco.dll
FAILED: bin/mujoco.dll lib/mujoco.lib 
cmd.exe /C "cd . && C:\bld\mujoco_1653515866886\_build_env\Library\bin\cmake.exe -E vs_link_dll --intdir=CMakeFiles\mujoco.dir --rc=C:\PROGRA~2\WI3CF2~1\10\bin\100220~1.0\x64\rc.exe --mt=C:\PROGRA~2\WI3CF2~1\10\bin\100220~1.0\x64\mt.exe --manifests  -- C:\PROGRA~1\LLVM\bin\lld-link.exe /nologo CMakeFiles\mujoco.dir\src\engine\engine_callback.c.obj CMakeFiles\mujoco.dir\src\engine\engine_collision_box.c.obj CMakeFiles\mujoco.dir\src\engine\engine_collision_convex.c.obj CMakeFiles\mujoco.dir\src\engine\engine_collision_driver.c.obj CMakeFiles\mujoco.dir\src\engine\engine_collision_primitive.c.obj CMakeFiles\mujoco.dir\src\engine\engine_core_constraint.c.obj CMakeFiles\mujoco.dir\src\engine\engine_core_smooth.c.obj CMakeFiles\mujoco.dir\src\engine\engine_derivative.c.obj CMakeFiles\mujoco.dir\src\engine\engine_file.c.obj CMakeFiles\mujoco.dir\src\engine\engine_forward.c.obj CMakeFiles\mujoco.dir\src\engine\engine_inverse.c.obj CMakeFiles\mujoco.dir\src\engine\engine_io.c.obj CMakeFiles\mujoco.dir\src\engine\engine_print.c.obj CMakeFiles\mujoco.dir\src\engine\engine_ray.c.obj CMakeFiles\mujoco.dir\src\engine\engine_sensor.c.obj CMakeFiles\mujoco.dir\src\engine\engine_setconst.c.obj CMakeFiles\mujoco.dir\src\engine\engine_solver.c.obj CMakeFiles\mujoco.dir\src\engine\engine_support.c.obj CMakeFiles\mujoco.dir\src\engine\engine_util_blas.c.obj CMakeFiles\mujoco.dir\src\engine\engine_util_errmem.c.obj CMakeFiles\mujoco.dir\src\engine\engine_util_misc.c.obj CMakeFiles\mujoco.dir\src\engine\engine_util_solve.c.obj CMakeFiles\mujoco.dir\src\engine\engine_util_sparse.c.obj CMakeFiles\mujoco.dir\src\engine\engine_util_spatial.c.obj CMakeFiles\mujoco.dir\src\engine\engine_vfs.c.obj CMakeFiles\mujoco.dir\src\engine\engine_vis_init.c.obj CMakeFiles\mujoco.dir\src\engine\engine_vis_interact.c.obj CMakeFiles\mujoco.dir\src\engine\engine_vis_visualize.c.obj CMakeFiles\mujoco.dir\src\user\user_composite.cc.obj CMakeFiles\mujoco.dir\src\user\user_mesh.cc.obj CMakeFiles\mujoco.dir\src\user\user_model.cc.obj CMakeFiles\mujoco.dir\src\user\user_objects.cc.obj CMakeFiles\mujoco.dir\src\user\user_util.cc.obj CMakeFiles\mujoco.dir\src\xml\xml_api.cc.obj CMakeFiles\mujoco.dir\src\xml\xml_base.cc.obj CMakeFiles\mujoco.dir\src\xml\xml.cc.obj CMakeFiles\mujoco.dir\src\xml\xml_native_reader.cc.obj CMakeFiles\mujoco.dir\src\xml\xml_numeric_format.cc.obj CMakeFiles\mujoco.dir\src\xml\xml_native_writer.cc.obj CMakeFiles\mujoco.dir\src\xml\xml_urdf.cc.obj CMakeFiles\mujoco.dir\src\xml\xml_util.cc.obj CMakeFiles\mujoco.dir\src\render\glad\glad.c.obj CMakeFiles\mujoco.dir\src\render\glad\loader.cc.obj CMakeFiles\mujoco.dir\src\render\render_context.c.obj CMakeFiles\mujoco.dir\src\render\render_gl2.c.obj CMakeFiles\mujoco.dir\src\render\render_gl3.c.obj CMakeFiles\mujoco.dir\src\render\render_util.c.obj CMakeFiles\mujoco.dir\src\ui\ui_main.c.obj  /out:bin\mujoco.dll /implib:lib\mujoco.lib /pdb:bin\mujoco.pdb /dll /version:2.2 /machine:x64 /INCREMENTAL:NO  /OPT:REF /OPT:ICF=5  C:\bld\mujoco_1653515866886\_h_env\Library\lib\ccd.lib  lib\lodepng.lib  lib\tinyobjloader.lib  tinyxml2.lib  C:\bld\mujoco_1653515866886\_h_env\Library\lib\qhull_r.lib  kernel32.lib user32.lib gdi32.lib winspool.lib shell32.lib ole32.lib oleaut32.lib uuid.lib comdlg32.lib advapi32.lib  && cd ."
LINK: command "C:\PROGRA~1\LLVM\bin\lld-link.exe /nologo CMakeFiles\mujoco.dir\src\engine\engine_callback.c.obj CMakeFiles\mujoco.dir\src\engine\engine_collision_box.c.obj CMakeFiles\mujoco.dir\src\engine\engine_collision_convex.c.obj CMakeFiles\mujoco.dir\src\engine\engine_collision_driver.c.obj CMakeFiles\mujoco.dir\src\engine\engine_collision_primitive.c.obj CMakeFiles\mujoco.dir\src\engine\engine_core_constraint.c.obj CMakeFiles\mujoco.dir\src\engine\engine_core_smooth.c.obj CMakeFiles\mujoco.dir\src\engine\engine_derivative.c.obj CMakeFiles\mujoco.dir\src\engine\engine_file.c.obj CMakeFiles\mujoco.dir\src\engine\engine_forward.c.obj CMakeFiles\mujoco.dir\src\engine\engine_inverse.c.obj CMakeFiles\mujoco.dir\src\engine\engine_io.c.obj CMakeFiles\mujoco.dir\src\engine\engine_print.c.obj CMakeFiles\mujoco.dir\src\engine\engine_ray.c.obj CMakeFiles\mujoco.dir\src\engine\engine_sensor.c.obj CMakeFiles\mujoco.dir\src\engine\engine_setconst.c.obj CMakeFiles\mujoco.dir\src\engine\engine_solver.c.obj CMakeFiles\mujoco.dir\src\engine\engine_support.c.obj CMakeFiles\mujoco.dir\src\engine\engine_util_blas.c.obj CMakeFiles\mujoco.dir\src\engine\engine_util_errmem.c.obj CMakeFiles\mujoco.dir\src\engine\engine_util_misc.c.obj CMakeFiles\mujoco.dir\src\engine\engine_util_solve.c.obj CMakeFiles\mujoco.dir\src\engine\engine_util_sparse.c.obj CMakeFiles\mujoco.dir\src\engine\engine_util_spatial.c.obj CMakeFiles\mujoco.dir\src\engine\engine_vfs.c.obj CMakeFiles\mujoco.dir\src\engine\engine_vis_init.c.obj CMakeFiles\mujoco.dir\src\engine\engine_vis_interact.c.obj CMakeFiles\mujoco.dir\src\engine\engine_vis_visualize.c.obj CMakeFiles\mujoco.dir\src\user\user_composite.cc.obj CMakeFiles\mujoco.dir\src\user\user_mesh.cc.obj CMakeFiles\mujoco.dir\src\user\user_model.cc.obj CMakeFiles\mujoco.dir\src\user\user_objects.cc.obj CMakeFiles\mujoco.dir\src\user\user_util.cc.obj CMakeFiles\mujoco.dir\src\xml\xml_api.cc.obj CMakeFiles\mujoco.dir\src\xml\xml_base.cc.obj CMakeFiles\mujoco.dir\src\xml\xml.cc.obj CMakeFiles\mujoco.dir\src\xml\xml_native_reader.cc.obj CMakeFiles\mujoco.dir\src\xml\xml_numeric_format.cc.obj CMakeFiles\mujoco.dir\src\xml\xml_native_writer.cc.obj CMakeFiles\mujoco.dir\src\xml\xml_urdf.cc.obj CMakeFiles\mujoco.dir\src\xml\xml_util.cc.obj CMakeFiles\mujoco.dir\src\render\glad\glad.c.obj CMakeFiles\mujoco.dir\src\render\glad\loader.cc.obj CMakeFiles\mujoco.dir\src\render\render_context.c.obj CMakeFiles\mujoco.dir\src\render\render_gl2.c.obj CMakeFiles\mujoco.dir\src\render\render_gl3.c.obj CMakeFiles\mujoco.dir\src\render\render_util.c.obj CMakeFiles\mujoco.dir\src\ui\ui_main.c.obj /out:bin\mujoco.dll /implib:lib\mujoco.lib /pdb:bin\mujoco.pdb /dll /version:2.2 /machine:x64 /INCREMENTAL:NO /OPT:REF /OPT:ICF=5 C:\bld\mujoco_1653515866886\_h_env\Library\lib\ccd.lib lib\lodepng.lib lib\tinyobjloader.lib tinyxml2.lib C:\bld\mujoco_1653515866886\_h_env\Library\lib\qhull_r.lib kernel32.lib user32.lib gdi32.lib winspool.lib shell32.lib ole32.lib oleaut32.lib uuid.lib comdlg32.lib advapi32.lib /MANIFEST /MANIFESTFILE:bin\mujoco.dll.manifest" failed (exit code 1) with the following output:
lld-link: error: undefined symbol: ccd_vec3_origin

>>> referenced by C:\bld\mujoco_1653515866886\work\src\engine\engine_collision_convex.c

>>>               CMakeFiles\mujoco.dir\src\engine\engine_collision_convex.c.obj

ninja: build stopped: subcommand failed.
Traceback (most recent call last):
~~~

saran-t commented:
You cannot devendor `glad`. This is a heavily modified, generated code. It must be treated as part of our codebase.

traversaro commented:
> You cannot devendor `glad`. This is a heavily modified, generated code. It must be treated as part of our codebase.

Ack, I did not noticed that the symbols are also modified to avoid conflicts. I modified the bullet point to address that, thanks!

traversaro commented:
> > Understand if it is a problem that we use libccd compiled with float support while mujoco expect libccd with double support
> 
> On this, there are quite a scary `Wincompatible-pointer-types` warnings during compilation. At a first glance I would expect nothing would work, but apparently tests are passing. Perhaps it is just a lucky (or unlucky, in a sense) case.
> Warnings from libccd use

For this, we probably will swithc libccd to use double as discussed in https://github.com/conda-forge/libccd-feedstock/issues/4 and proposed in https://github.com/conda-forge/libccd-feedstock/issues/5 .

traversaro commented:
PRs that are prerequisite for this PR:
* https://github.com/conda-forge/staged-recipes/pull/19104
* https://github.com/conda-forge/tinyobjloader-feedstock/pull/2
* https://github.com/conda-forge/libccd-feedstock/pull/5

traversaro commented:
>  Get Windows to work (libccd seems to be missing some symbols)

It seems that the problem is still there even with `ccd-double`:
~~~
[51/51] Linking CXX shared library bin\mujoco.dll
FAILED: bin/mujoco.dll lib/mujoco.lib 
cmd.exe /C "cd . && C:\bld\mujoco_1654716236692\_build_env\Library\bin\cmake.exe -E vs_link_dll --intdir=CMakeFiles\mujoco.dir --rc=C:\PROGRA~2\WI3CF2~1\10\bin\100220~1.0\x64\rc.exe --mt=C:\PROGRA~2\WI3CF2~1\10\bin\100220~1.0\x64\mt.exe --manifests  -- C:\PROGRA~1\LLVM\bin\lld-link.exe /nologo CMakeFiles\mujoco.dir\src\engine\engine_callback.c.obj CMakeFiles\mujoco.dir\src\engine\engine_collision_box.c.obj CMakeFiles\mujoco.dir\src\engine\engine_collision_convex.c.obj CMakeFiles\mujoco.dir\src\engine\engine_collision_driver.c.obj CMakeFiles\mujoco.dir\src\engine\engine_collision_primitive.c.obj CMakeFiles\mujoco.dir\src\engine\engine_core_constraint.c.obj CMakeFiles\mujoco.dir\src\engine\engine_core_smooth.c.obj CMakeFiles\mujoco.dir\src\engine\engine_derivative.c.obj CMakeFiles\mujoco.dir\src\engine\engine_file.c.obj CMakeFiles\mujoco.dir\src\engine\engine_forward.c.obj CMakeFiles\mujoco.dir\src\engine\engine_inverse.c.obj CMakeFiles\mujoco.dir\src\engine\engine_io.c.obj CMakeFiles\mujoco.dir\src\engine\engine_print.c.obj CMakeFiles\mujoco.dir\src\engine\engine_ray.c.obj CMakeFiles\mujoco.dir\src\engine\engine_sensor.c.obj CMakeFiles\mujoco.dir\src\engine\engine_setconst.c.obj CMakeFiles\mujoco.dir\src\engine\engine_solver.c.obj CMakeFiles\mujoco.dir\src\engine\engine_support.c.obj CMakeFiles\mujoco.dir\src\engine\engine_util_blas.c.obj CMakeFiles\mujoco.dir\src\engine\engine_util_errmem.c.obj CMakeFiles\mujoco.dir\src\engine\engine_util_misc.c.obj CMakeFiles\mujoco.dir\src\engine\engine_util_solve.c.obj CMakeFiles\mujoco.dir\src\engine\engine_util_sparse.c.obj CMakeFiles\mujoco.dir\src\engine\engine_util_spatial.c.obj CMakeFiles\mujoco.dir\src\engine\engine_vfs.c.obj CMakeFiles\mujoco.dir\src\engine\engine_vis_init.c.obj CMakeFiles\mujoco.dir\src\engine\engine_vis_interact.c.obj CMakeFiles\mujoco.dir\src\engine\engine_vis_visualize.c.obj CMakeFiles\mujoco.dir\src\user\user_composite.cc.obj CMakeFiles\mujoco.dir\src\user\user_mesh.cc.obj CMakeFiles\mujoco.dir\src\user\user_model.cc.obj CMakeFiles\mujoco.dir\src\user\user_objects.cc.obj CMakeFiles\mujoco.dir\src\user\user_util.cc.obj CMakeFiles\mujoco.dir\src\xml\xml_api.cc.obj CMakeFiles\mujoco.dir\src\xml\xml_base.cc.obj CMakeFiles\mujoco.dir\src\xml\xml.cc.obj CMakeFiles\mujoco.dir\src\xml\xml_native_reader.cc.obj CMakeFiles\mujoco.dir\src\xml\xml_numeric_format.cc.obj CMakeFiles\mujoco.dir\src\xml\xml_native_writer.cc.obj CMakeFiles\mujoco.dir\src\xml\xml_urdf.cc.obj CMakeFiles\mujoco.dir\src\xml\xml_util.cc.obj CMakeFiles\mujoco.dir\src\render\glad\glad.c.obj CMakeFiles\mujoco.dir\src\render\glad\loader.cc.obj CMakeFiles\mujoco.dir\src\render\render_context.c.obj CMakeFiles\mujoco.dir\src\render\render_gl2.c.obj CMakeFiles\mujoco.dir\src\render\render_gl3.c.obj CMakeFiles\mujoco.dir\src\render\render_util.c.obj CMakeFiles\mujoco.dir\src\ui\ui_main.c.obj  /out:bin\mujoco.dll /implib:lib\mujoco.lib /pdb:bin\mujoco.pdb /dll /version:2.2 /machine:x64 /INCREMENTAL:NO  /OPT:REF /OPT:ICF=5  C:\bld\mujoco_1654716236692\_h_env\Library\lib\ccd.lib  C:\bld\mujoco_1654716236692\_h_env\Library\lib\lodepng.lib  lib\tinyobjloader.lib  tinyxml2.lib  C:\bld\mujoco_1654716236692\_h_env\Library\lib\qhull_r.lib  kernel32.lib user32.lib gdi32.lib winspool.lib shell32.lib ole32.lib oleaut32.lib uuid.lib comdlg32.lib advapi32.lib  && cd ."
LINK: command "C:\PROGRA~1\LLVM\bin\lld-link.exe /nologo CMakeFiles\mujoco.dir\src\engine\engine_callback.c.obj CMakeFiles\mujoco.dir\src\engine\engine_collision_box.c.obj CMakeFiles\mujoco.dir\src\engine\engine_collision_convex.c.obj CMakeFiles\mujoco.dir\src\engine\engine_collision_driver.c.obj CMakeFiles\mujoco.dir\src\engine\engine_collision_primitive.c.obj CMakeFiles\mujoco.dir\src\engine\engine_core_constraint.c.obj CMakeFiles\mujoco.dir\src\engine\engine_core_smooth.c.obj CMakeFiles\mujoco.dir\src\engine\engine_derivative.c.obj CMakeFiles\mujoco.dir\src\engine\engine_file.c.obj CMakeFiles\mujoco.dir\src\engine\engine_forward.c.obj CMakeFiles\mujoco.dir\src\engine\engine_inverse.c.obj CMakeFiles\mujoco.dir\src\engine\engine_io.c.obj CMakeFiles\mujoco.dir\src\engine\engine_print.c.obj CMakeFiles\mujoco.dir\src\engine\engine_ray.c.obj CMakeFiles\mujoco.dir\src\engine\engine_sensor.c.obj CMakeFiles\mujoco.dir\src\engine\engine_setconst.c.obj CMakeFiles\mujoco.dir\src\engine\engine_solver.c.obj CMakeFiles\mujoco.dir\src\engine\engine_support.c.obj CMakeFiles\mujoco.dir\src\engine\engine_util_blas.c.obj CMakeFiles\mujoco.dir\src\engine\engine_util_errmem.c.obj CMakeFiles\mujoco.dir\src\engine\engine_util_misc.c.obj CMakeFiles\mujoco.dir\src\engine\engine_util_solve.c.obj CMakeFiles\mujoco.dir\src\engine\engine_util_sparse.c.obj CMakeFiles\mujoco.dir\src\engine\engine_util_spatial.c.obj CMakeFiles\mujoco.dir\src\engine\engine_vfs.c.obj CMakeFiles\mujoco.dir\src\engine\engine_vis_init.c.obj CMakeFiles\mujoco.dir\src\engine\engine_vis_interact.c.obj CMakeFiles\mujoco.dir\src\engine\engine_vis_visualize.c.obj CMakeFiles\mujoco.dir\src\user\user_composite.cc.obj CMakeFiles\mujoco.dir\src\user\user_mesh.cc.obj CMakeFiles\mujoco.dir\src\user\user_model.cc.obj CMakeFiles\mujoco.dir\src\user\user_objects.cc.obj CMakeFiles\mujoco.dir\src\user\user_util.cc.obj CMakeFiles\mujoco.dir\src\xml\xml_api.cc.obj CMakeFiles\mujoco.dir\src\xml\xml_base.cc.obj CMakeFiles\mujoco.dir\src\xml\xml.cc.obj CMakeFiles\mujoco.dir\src\xml\xml_native_reader.cc.obj CMakeFiles\mujoco.dir\src\xml\xml_numeric_format.cc.obj CMakeFiles\mujoco.dir\src\xml\xml_native_writer.cc.obj CMakeFiles\mujoco.dir\src\xml\xml_urdf.cc.obj CMakeFiles\mujoco.dir\src\xml\xml_util.cc.obj CMakeFiles\mujoco.dir\src\render\glad\glad.c.obj CMakeFiles\mujoco.dir\src\render\glad\loader.cc.obj CMakeFiles\mujoco.dir\src\render\render_context.c.obj CMakeFiles\mujoco.dir\src\render\render_gl2.c.obj CMakeFiles\mujoco.dir\src\render\render_gl3.c.obj CMakeFiles\mujoco.dir\src\render\render_util.c.obj CMakeFiles\mujoco.dir\src\ui\ui_main.c.obj /out:bin\mujoco.dll /implib:lib\mujoco.lib /pdb:bin\mujoco.pdb /dll /version:2.2 /machine:x64 /INCREMENTAL:NO /OPT:REF /OPT:ICF=5 C:\bld\mujoco_1654716236692\_h_env\Library\lib\ccd.lib C:\bld\mujoco_1654716236692\_h_env\Library\lib\lodepng.lib lib\tinyobjloader.lib tinyxml2.lib C:\bld\mujoco_1654716236692\_h_env\Library\lib\qhull_r.lib kernel32.lib user32.lib gdi32.lib winspool.lib shell32.lib ole32.lib oleaut32.lib uuid.lib comdlg32.lib advapi32.lib /MANIFEST /MANIFESTFILE:bin\mujoco.dll.manifest" failed (exit code 1) with the following output:
lld-link: error: undefined symbol: ccd_vec3_origin
>>> referenced by C:\bld\mujoco_1654716236692\work\src\engine\engine_collision_convex.c

>>>               CMakeFiles\mujoco.dir\src\engine\engine_collision_convex.c.obj

ninja: build stopped: subcommand failed.
~~~

traversaro commented:
The symbol seems to be in the library:
~~~
(libccd) C:\Users\STraversaro\AppData\Local\mambaforge\envs\libccd\Library\lib>dumpbin /exports ccd.lib
Microsoft (R) COFF/PE Dumper Version 14.16.27048.0
Copyright (C) Microsoft Corporation.  All rights reserved.


Dump of file ccd.lib

File Type: LIBRARY

     Exports

       ordinal    name

                  __ccdSupport
                  ccdFirstDirDefault
                  ccdGJKIntersect
                  ccdGJKPenetration
                  ccdGJKSeparate
                  ccdMPRIntersect
                  ccdMPRPenetration
                  ccdPtAddEdge
                  ccdPtAddFace
                  ccdPtAddVertex
                  ccdPtDestroy
                  ccdPtDumpSVT
                  ccdPtDumpSVT2
                  ccdPtInit
                  ccdPtNearest
                  ccdPtRecomputeDistances
                  ccdVec3PointSegmentDist2
                  ccdVec3PointTriDist2
                  ccd_points_on_sphere
                  ccd_points_on_sphere_len
                  ccd_vec3_origin

  Summary

          B7 .debug$S
          14 .idata$2
          14 .idata$3
           8 .idata$4
           8 .idata$5
           8 .idata$6
~~~

So the problem is probably elsewhere.

traversaro commented:
> > Get Windows to work (libccd seems to be missing some symbols)
> 
> It seems that the problem is still there even with `ccd-double`:
> 
> ```
> [51/51] Linking CXX shared library bin\mujoco.dll
> FAILED: bin/mujoco.dll lib/mujoco.lib 
> cmd.exe /C "cd . && C:\bld\mujoco_1654716236692\_build_env\Library\bin\cmake.exe -E vs_link_dll --intdir=CMakeFiles\mujoco.dir --rc=C:\PROGRA~2\WI3CF2~1\10\bin\100220~1.0\x64\rc.exe --mt=C:\PROGRA~2\WI3CF2~1\10\bin\100220~1.0\x64\mt.exe --manifests  -- C:\PROGRA~1\LLVM\bin\lld-link.exe /nologo CMakeFiles\mujoco.dir\src\engine\engine_callback.c.obj CMakeFiles\mujoco.dir\src\engine\engine_collision_box.c.obj CMakeFiles\mujoco.dir\src\engine\engine_collision_convex.c.obj CMakeFiles\mujoco.dir\src\engine\engine_collision_driver.c.obj CMakeFiles\mujoco.dir\src\engine\engine_collision_primitive.c.obj CMakeFiles\mujoco.dir\src\engine\engine_core_constraint.c.obj CMakeFiles\mujoco.dir\src\engine\engine_core_smooth.c.obj CMakeFiles\mujoco.dir\src\engine\engine_derivative.c.obj CMakeFiles\mujoco.dir\src\engine\engine_file.c.obj CMakeFiles\mujoco.dir\src\engine\engine_forward.c.obj CMakeFiles\mujoco.dir\src\engine\engine_inverse.c.obj CMakeFiles\mujoco.dir\src\engine\engine_io.c.obj CMakeFiles\mujoco.dir\src\engine\engine_print.c.obj CMakeFiles\mujoco.dir\src\engine\engine_ray.c.obj CMakeFiles\mujoco.dir\src\engine\engine_sensor.c.obj CMakeFiles\mujoco.dir\src\engine\engine_setconst.c.obj CMakeFiles\mujoco.dir\src\engine\engine_solver.c.obj CMakeFiles\mujoco.dir\src\engine\engine_support.c.obj CMakeFiles\mujoco.dir\src\engine\engine_util_blas.c.obj CMakeFiles\mujoco.dir\src\engine\engine_util_errmem.c.obj CMakeFiles\mujoco.dir\src\engine\engine_util_misc.c.obj CMakeFiles\mujoco.dir\src\engine\engine_util_solve.c.obj CMakeFiles\mujoco.dir\src\engine\engine_util_sparse.c.obj CMakeFiles\mujoco.dir\src\engine\engine_util_spatial.c.obj CMakeFiles\mujoco.dir\src\engine\engine_vfs.c.obj CMakeFiles\mujoco.dir\src\engine\engine_vis_init.c.obj CMakeFiles\mujoco.dir\src\engine\engine_vis_interact.c.obj CMakeFiles\mujoco.dir\src\engine\engine_vis_visualize.c.obj CMakeFiles\mujoco.dir\src\user\user_composite.cc.obj CMakeFiles\mujoco.dir\src\user\user_mesh.cc.obj CMakeFiles\mujoco.dir\src\user\user_model.cc.obj CMakeFiles\mujoco.dir\src\user\user_objects.cc.obj CMakeFiles\mujoco.dir\src\user\user_util.cc.obj CMakeFiles\mujoco.dir\src\xml\xml_api.cc.obj CMakeFiles\mujoco.dir\src\xml\xml_base.cc.obj CMakeFiles\mujoco.dir\src\xml\xml.cc.obj CMakeFiles\mujoco.dir\src\xml\xml_native_reader.cc.obj CMakeFiles\mujoco.dir\src\xml\xml_numeric_format.cc.obj CMakeFiles\mujoco.dir\src\xml\xml_native_writer.cc.obj CMakeFiles\mujoco.dir\src\xml\xml_urdf.cc.obj CMakeFiles\mujoco.dir\src\xml\xml_util.cc.obj CMakeFiles\mujoco.dir\src\render\glad\glad.c.obj CMakeFiles\mujoco.dir\src\render\glad\loader.cc.obj CMakeFiles\mujoco.dir\src\render\render_context.c.obj CMakeFiles\mujoco.dir\src\render\render_gl2.c.obj CMakeFiles\mujoco.dir\src\render\render_gl3.c.obj CMakeFiles\mujoco.dir\src\render\render_util.c.obj CMakeFiles\mujoco.dir\src\ui\ui_main.c.obj  /out:bin\mujoco.dll /implib:lib\mujoco.lib /pdb:bin\mujoco.pdb /dll /version:2.2 /machine:x64 /INCREMENTAL:NO  /OPT:REF /OPT:ICF=5  C:\bld\mujoco_1654716236692\_h_env\Library\lib\ccd.lib  C:\bld\mujoco_1654716236692\_h_env\Library\lib\lodepng.lib  lib\tinyobjloader.lib  tinyxml2.lib  C:\bld\mujoco_1654716236692\_h_env\Library\lib\qhull_r.lib  kernel32.lib user32.lib gdi32.lib winspool.lib shell32.lib ole32.lib oleaut32.lib uuid.lib comdlg32.lib advapi32.lib  && cd ."
> LINK: command "C:\PROGRA~1\LLVM\bin\lld-link.exe /nologo CMakeFiles\mujoco.dir\src\engine\engine_callback.c.obj CMakeFiles\mujoco.dir\src\engine\engine_collision_box.c.obj CMakeFiles\mujoco.dir\src\engine\engine_collision_convex.c.obj CMakeFiles\mujoco.dir\src\engine\engine_collision_driver.c.obj CMakeFiles\mujoco.dir\src\engine\engine_collision_primitive.c.obj CMakeFiles\mujoco.dir\src\engine\engine_core_constraint.c.obj CMakeFiles\mujoco.dir\src\engine\engine_core_smooth.c.obj CMakeFiles\mujoco.dir\src\engine\engine_derivative.c.obj CMakeFiles\mujoco.dir\src\engine\engine_file.c.obj CMakeFiles\mujoco.dir\src\engine\engine_forward.c.obj CMakeFiles\mujoco.dir\src\engine\engine_inverse.c.obj CMakeFiles\mujoco.dir\src\engine\engine_io.c.obj CMakeFiles\mujoco.dir\src\engine\engine_print.c.obj CMakeFiles\mujoco.dir\src\engine\engine_ray.c.obj CMakeFiles\mujoco.dir\src\engine\engine_sensor.c.obj CMakeFiles\mujoco.dir\src\engine\engine_setconst.c.obj CMakeFiles\mujoco.dir\src\engine\engine_solver.c.obj CMakeFiles\mujoco.dir\src\engine\engine_support.c.obj CMakeFiles\mujoco.dir\src\engine\engine_util_blas.c.obj CMakeFiles\mujoco.dir\src\engine\engine_util_errmem.c.obj CMakeFiles\mujoco.dir\src\engine\engine_util_misc.c.obj CMakeFiles\mujoco.dir\src\engine\engine_util_solve.c.obj CMakeFiles\mujoco.dir\src\engine\engine_util_sparse.c.obj CMakeFiles\mujoco.dir\src\engine\engine_util_spatial.c.obj CMakeFiles\mujoco.dir\src\engine\engine_vfs.c.obj CMakeFiles\mujoco.dir\src\engine\engine_vis_init.c.obj CMakeFiles\mujoco.dir\src\engine\engine_vis_interact.c.obj CMakeFiles\mujoco.dir\src\engine\engine_vis_visualize.c.obj CMakeFiles\mujoco.dir\src\user\user_composite.cc.obj CMakeFiles\mujoco.dir\src\user\user_mesh.cc.obj CMakeFiles\mujoco.dir\src\user\user_model.cc.obj CMakeFiles\mujoco.dir\src\user\user_objects.cc.obj CMakeFiles\mujoco.dir\src\user\user_util.cc.obj CMakeFiles\mujoco.dir\src\xml\xml_api.cc.obj CMakeFiles\mujoco.dir\src\xml\xml_base.cc.obj CMakeFiles\mujoco.dir\src\xml\xml.cc.obj CMakeFiles\mujoco.dir\src\xml\xml_native_reader.cc.obj CMakeFiles\mujoco.dir\src\xml\xml_numeric_format.cc.obj CMakeFiles\mujoco.dir\src\xml\xml_native_writer.cc.obj CMakeFiles\mujoco.dir\src\xml\xml_urdf.cc.obj CMakeFiles\mujoco.dir\src\xml\xml_util.cc.obj CMakeFiles\mujoco.dir\src\render\glad\glad.c.obj CMakeFiles\mujoco.dir\src\render\glad\loader.cc.obj CMakeFiles\mujoco.dir\src\render\render_context.c.obj CMakeFiles\mujoco.dir\src\render\render_gl2.c.obj CMakeFiles\mujoco.dir\src\render\render_gl3.c.obj CMakeFiles\mujoco.dir\src\render\render_util.c.obj CMakeFiles\mujoco.dir\src\ui\ui_main.c.obj /out:bin\mujoco.dll /implib:lib\mujoco.lib /pdb:bin\mujoco.pdb /dll /version:2.2 /machine:x64 /INCREMENTAL:NO /OPT:REF /OPT:ICF=5 C:\bld\mujoco_1654716236692\_h_env\Library\lib\ccd.lib C:\bld\mujoco_1654716236692\_h_env\Library\lib\lodepng.lib lib\tinyobjloader.lib tinyxml2.lib C:\bld\mujoco_1654716236692\_h_env\Library\lib\qhull_r.lib kernel32.lib user32.lib gdi32.lib winspool.lib shell32.lib ole32.lib oleaut32.lib uuid.lib comdlg32.lib advapi32.lib /MANIFEST /MANIFESTFILE:bin\mujoco.dll.manifest" failed (exit code 1) with the following output:
> lld-link: error: undefined symbol: ccd_vec3_origin
> >>> referenced by C:\bld\mujoco_1654716236692\work\src\engine\engine_collision_convex.c
> 
> >>>               CMakeFiles\mujoco.dir\src\engine\engine_collision_convex.c.obj
> 
> ninja: build stopped: subcommand failed.
> ```

I was unable to solve the problem. However, as that symbol is used just to access a `0.0 0.0 0.0` vector, I patched the source code to avoid using it and instead rely on a zero vector contained in a local instance: https://github.com/conda-forge/staged-recipes/pull/19049/commits/6e083278aebf00b50d351e8b86184b6031b4646b .

traversaro commented:
Python bindings on Windows are failing with:
~~~
  -- Building for: NMake Makefiles
  CMake Error at CMakeLists.txt:15 (project):
    Generator
  
      NMake Makefiles
  
    does not support platform specification, but platform
  
      x64
  
    was specified.
  
  
  CMake Error: CMAKE_C_COMPILER not set, after EnableLanguage
  CMake Error: CMAKE_CXX_COMPILER not set, after EnableLanguage
~~~

traversaro commented:
The recipe is ready for review @conda-forge/help-c-cpp @conda-forge/help-python-c .

Some additional notes/points that is better to explain before:
* AVX is currently disabled due to https://github.com/conda-forge/conda-forge.github.io/issues/1261 .
* Currently [`tinyobjloader`](https://github.com/tinyobjloader/tinyobjloader) is not de-vendored as `mujoco` depends on a feature/API that is part of v2 (in rc since three years, see https://github.com/tinyobjloader/tinyobjloader/tags), and not part of v1 that is the version available in conda-forge: https://github.com/conda-forge/tinyobjloader-feedstock. However, the symbols of internal libraries are hidden (see https://github.com/deepmind/mujoco/blob/64bc6d27b242d5c5683b4a03dd36f1844b910742/cmake/MujocoOptions.cmake#L53), so I think it is safe to merge it as it is and switch to de-vendor tinyobjloader as soon as v2 is available on conda-forge.
* macOS build is disabled as the library requires a new macOS SKD to be used (see https://conda-forge.org/docs/maintainer/knowledge_base.html#requiring-newer-macos-sdks). So the idea is to try to enable in the feedstock (where a `recipe/conda_build_config.yaml` would be considered) once the recipe is merged.

carterbox reviewed (changes requested):
In regards to revendoring tinyobjlibrary. I believe this is OK because the release candidate version should be published on a conda-forge rc channel which means that would probably require end users to add a special channel in order for the package solver to find a solution. I have questions though:

1. tinyobjlibrary is statically linked into the mujoco shared library?
2. Where is the license for tinyobjlibrary since it is being statically linked into mujoco?

traversaro commented on recipes/mujoco/meta.yaml:
```diff
@@ -0,0 +1,127 @@
+      # abseil-cpp is used only for tests, we can 
+      # safely ignore its run_exports
+      ignore_run_exports:
+        - abseil-cpp
```
On Python (differently from C++) the library is used in the bindings themselfs, not on tests, see for example the Linux logs:
~~~
2022-06-18T14:40:23.0083572Z    INFO (mujoco-python,lib/python3.8/site-packages/mujoco/_structs.cpython-38-x86_64-linux-gnu.so): Needed DSO lib/libabsl_hash.so.2103.0.1 found in conda-forge::abseil-cpp-20210324.2-h9c3ff4c_0
2022-06-18T14:40:23.0147391Z    INFO (mujoco-python,lib/python3.8/site-packages/mujoco/_structs.cpython-38-x86_64-linux-gnu.so): Needed DSO lib/libabsl_raw_hash_set.so.2103.0.1 found in conda-forge::abseil-cpp-20210324.2-h9c3ff4c_0
2022-06-18T14:40:23.0176888Z    INFO (mujoco-python,lib/python3.8/site-packages/mujoco/_structs.cpython-38-x86_64-linux-gnu.so): Needed DSO x86_64-conda-linux-gnu/sysroot/lib64/libpthread.so.0 found in CDT/compiler package conda-forge::sysroot_linux-64-2.12-he073ed8_15
2022-06-18T14:40:23.0238563Z    INFO (mujoco-python,lib/python3.8/site-packages/mujoco/_structs.cpython-38-x86_64-linux-gnu.so): Needed DSO lib/libmujoco.so.2.2.0 found in home/conda/staged-recipes/build_artifacts::mujoco-cxx-2.2.0-h696428e_0
2022-06-18T14:40:23.0298191Z    INFO (mujoco-python,lib/python3.8/site-packages/mujoco/_structs.cpython-38-x86_64-linux-gnu.so): Needed DSO lib/libabsl_throw_delegate.so.2103.0.1 found in conda-forge::abseil-cpp-20210324.2-h9c3ff4c_0
~~~

This is the reason why I did not check for that warning on Windows. However, the Windows warning is indeed fishy. By quickly looking at the abseil Windows package, in particular the size of the shipped `.lib`, at a first glance it seems that a library .lib is shipped, instead of a import library, however I need to check better.

traversaro replied:
Ok, I checked a bit the situation. abseil-cpp is still pinned to a version that shipped static libraries on Windows (i.e. before https://github.com/conda-forge/abseil-cpp-feedstock/pull/25), see https://github.com/conda-forge/conda-forge-pinning-feedstock/blob/a7759497a85d3c3a0e929583e5babacc922950c8/recipe/conda_build_config.yaml#L347 . Once an abseil with shared libraries is available, it will be used, and no modifications are required on the mujoco recipe. At this point, I think leaving avoiding to add the `ignore_run_exports` on the python side is probably the best/simplest thing to do.

carterbox replied:
Agreed. If a future version provides non-static linking it's fine.

traversaro commented:
> In regards to revendoring tinyobjlibrary. I believe this is OK because the release candidate version should be published on a conda-forge rc channel which means that would probably require end users to add a special channel in order for the package solver to find a solution. I have questions though:
> 
>     1. tinyobjlibrary is statically linked into the mujoco shared library?
> 
>     2. Where is the license for tinyobjlibrary since it is being statically linked into mujoco?

Yes, `tinyobjloader` is statically linked and I add its license in the `meta.yaml`.
